pkg/generators/extension.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generators
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/golang/glog"
+	"k8s.io/gengo/types"
+)
+
+// Extension tag to openapi extension
+var tagToExtension = map[string]string{
+	"patchMergeKey": "x-kubernetes-patch-merge-key",
+	"patchStrategy": "x-kubernetes-patch-strategy",
+	"listType":      "x-kubernetes-list-type",
+	"listTypeKey":   "x-kubernetes-list-type-keys",
+}
+
+// Enum values per extension
+var allowedExtensionValues = map[string][]string{
+	"x-kubernetes-patch-strategy": []string{"merge", "retainKeys"},
+	"x-kubernetes-list-type":      []string{"atomic", "set", "map"},
+}
+
+// TODO: Create interface for an extension validation function, and create
+// validation functions for different extensions.
+type Extension struct {
+	name   string   // Example: x-kubernetes-list-type
+	values []string // Example: atomic
+}
+
+func (e Extension) validateAllowedValues() error {
+	// allowedValues not set means no restrictions on values.
+	allowedValues, exists := allowedExtensionValues[e.name]
+	if !exists {
+		return nil
+	}
+	// For each extension value, validate that it is allowed.
+	for _, value := range e.values {
+		if !valueInArray(value, allowedValues) {
+			return fmt.Errorf("%s: value (%s) not allowed", e.name, value)
```

Comment:
Will users be able to figure out what source file/line produced this error? Please test the error output.

Also it's nice to list the allowed values.

Reply:
Error output is tested. Allowed values are now added to error.

It looks like incorporating the source file/line in the error would require some refactoring. I suggest we add that in a subsequent PR.